### DISCUSSION TITLE
MVP Progress

### DISCUSSION THREAD

**Discussion by KyleKing** [DesignPatternsAdventure/game#4] [Ideas]
- [x] Prototype code reload
- [x] Initial `core/*` and `task/*` layout
- [x] Copy over the resources we want from `community-rpg` and `kenny.ai` (@dimmerex)
- [x] Add the starting island and map tiles (@dimmerex in #8)
- [x] Implement arbiter logic for navigation to prevent walking on water (~~I'm thinking about using floats so that we can model sand and hills that might slow down a character~~ restore [community-rpg physics](https://github.com/DesignPatternsAdventure/community-rpg/blob/e2c726a421195f3ce7bf2725886590d1591d7ee1/src/rpg/views/game_view.py#L76-L79) @KyleKing)
- [x] Implement Views (~~Save menu?~~ Per comment below, saving will be automatic. Switch to view implemented in #6) (@KyleKing)
- [x] Reach >50% coverage with unit tests for core (@KyleKing)
- [x] Migrate Player Inventory from community-rpg and get pickaxe working again (@dimmerex in #11)
- [x] Restore camera from community-rpg (@KyleKing)
- [x] Migrate text notification and non-map UI elements (message box) from community-rpg (@dimmerex)
- [x] Migrate continuous save from community-rpg using the tilemap (@dimmerex)
- [x] Implement S task (@KyleKing)
- [x] Implement O task (blocked by pending refactor) (@KyleKing)
- [x] Create unit tests for S and O tasks (@KyleKing) - See [example arcade tests](https://github.com/pythonarcade/roguelike/blob/master/tests/test_game_window.py)
- [x] Plan LID tasks
- [ ] Refactor to integrate generic collision and nearby detection logic (Draft in #7) (@KyleKing) - See [notes on Tiled](https://api.arcade.academy/en/latest/examples/platform_tutorial/step_09.html) and consider [sprite following](https://api.arcade.academy/en/latest/examples/sprite_follow_simple.html)
- [x] Update README and switch to pft (#13) (@KyleKing)
- [x] Add demo gif to clarify that this is a code writing game
- [x] Move OS-specific instructions to `INSTALL.md` and make [default install process using `pyinstaller`](https://github.gatech.edu/CS6310-Group26/CS6310-Fall20-A6-26-Client) and an executable Python (@KyleKing)

**Comment by KyleKing** [comment]
**Persistent State Discussion**

I was thinking that we could utilize `sqlite` for persistence. We would have a set of main tables for game state (such as entity, sprite_state, etc.). The registration logic would know how to delegate this state and reload the `code_modules` using the `source_name` so that we don't need to pickle anything.

When we save a checkpoint, we can just duplicate the tables and rename based on the checkpoint time. Then we can have a `checkpoint` table which has columns `time, name, entity_table, sprite_state_table, etc.` so that we can display all possible checkpoints and restore from any of them

For packages, I was thinking [we could use `SQLModel`](https://github.com/tiangolo/sqlmodel) to turn the existing `BaseModels` into active records for the SQLite backend. Then we can use `alembic` to detect changes in the Models and auto-generate migrations for current and checkpointed tables so that any saved checkpoint can be restored in the most up-to-date version of the game code.

I think SQLite is a great option for performance and type support, but requiring database migrations would make add some complexity if users want to persist additional state beyond what is provided (although that shouldn't be necessary). Having a noSQL datastore wouldn't require migrations, but would be more difficult to ensure data quality (if going this route, we could consider [TinyDB](https://tinydb.readthedocs.io/en/latest/intro.html#why-use-tinydb) and duplicate JSON files on save). As an middle ground, we could [allow more flexibility by supporting JSON in SQLite](https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#sqlalchemy.dialects.sqlite.JSON)

For me, this is definitely a case of the hammer + nail problem, so @dimmerex feel free to comment with any issues I haven't considered or better approaches that you think of. I would be happy to consider any alternatives!

**Reply by dimmerex** [reply]
@KyleKing 
For the map, I think it makes sense to save it as JSON because the arcade [tilemap](https://api.arcade.academy/en/stable/api/tilemap.html#arcade-tilemap-tilemap) function takes in a JSON file.

For everything else, I actually tried out using sqlite last week, and it worked, but it just felt like overkill for what we were trying to do, mostly because most tables wouldn't be populated more than a few rows, unless we decide to keep a history of all the checkpoints for timestamps. But then the question is, do we need a history of checkpoints, or can we just keep the latest? Also pickle is not unusual to use by any means. There's plenty of games made in python that saves state using pickle.
